### PR TITLE
[FLINK-25857] Add committer metrics to track the status of committables - non-breaking

### PR DESCRIPTION
**Pull request by pvary**
## What is the purpose of the change

This is the a revisited non-breaking implementation of [FLIP-371: Provide initialization context for Committer creation in TwoPhaseCommittingSink](https://cwiki.apache.org/confluence/display/FLINK/FLIP-371%3A+Provide+initialization+context+for+Committer+creation+in+TwoPhaseCommittingSink).

Reverts the previous version [c3a07f98e5d1d7624adc967932f57d31355d9ddd](https://github.com/apache/flink/commit/c3a07f98e5d1d7624adc967932f57d31355d9ddd) with [50da9518e05bd0fe821f09e997582867bed5022a](https://github.com/apache/flink/pull/23876/commits/50da9518e05bd0fe821f09e997582867bed5022a) and [df9bd52593d33e0358eefd3008588084abe17b56](https://github.com/apache/flink/pull/23876/commits/df9bd52593d33e0358eefd3008588084abe17b56). Then adds the new implementation [a51eff4e2de3f80de61fb45fee81f0de68fd0af3](https://github.com/apache/flink/pull/23876/commits/a51eff4e2de3f80de61fb45fee81f0de68fd0af3) based on the [discussion](https://lists.apache.org/thread/h6nkgth838dlh5s90sd95zd6hlsxwx57) on the mailing list.

## Brief change log

API changes:
- `TwoPhaseCommittingSink` - new `createCommitter` method with default implementation and deprecation
- `CommitterInitContext` - new object for committer initialization and metrics
- `SinkCommitterMetricGroup` - new metrics group for committer related metrics
- `Sink.createWriter` - parameter changed to `WriterInitContext` with default implementation and deprecation
- `StatefulSink.createWriter` and `TwoPhaseCommittingSink.createWriter` method declarations are removed - and they fall back to the `Sink.createWriter`, which is backward compatible, since the return type is wider

Implementation:
- `InitContext` - new common context base class for `CommitterInitContext` and `WriterInitContext`
- `WriterInitContext` - replacement for `Sink.InitContext`
- `CommitterInitContext` - new context for committer creation
- `InitContextBase` - the common base for the `WriterInitContext` and `CommitterInitContext` implementation
- `Sink.createWriter(InitContext context)` is deprecated in favour of `createWriter(WriterInitContext context)` with default implementation
- `Sink.createWriter(WriterInitContext context)` is created with default implementation which wraps the `WriterInitContext` with `InitContextWrapper` to the old `Sink.InitContext` and calls the original method
- `StatefulSink.createWriter` and the `TwoPhaseCommitingSink.createWriter` methods are removed - they fall back to the `Sink.createWriter` method which is backward compatible
- `SinkWriterOperator` - to follow the `WriterInitContext` related changes
- `StatefulSinkWriterStateHandler` - Here we check if the writer created by `createWriter` is a `StatefulSinkWriter`
- `InternalSinkCommitterMetricGroup` - implementation for `SinkCommitterMetricGroup`
- `CommitterOperator` - to create the `CommitterInitContext` and `SinkCommitterMetricGroup`. And modify the initialization to  use the new `Committer` creation method, and create the metrics group
- Updating the metrics in the `CommitRequestImpl` and `CheckpointCommittableManagerImpl`
- Propagate the new metricGroup object to finally arrive to the `CommitRequestImpl` where we need it (`CommiterOperator` -> `CommittableCollector` -> `CheckpointCommittableManagerImpl` -> `SubtaskCommittableManager` -> `SubtaskCommittableManager` -> `CommitRequestImpl`)
- SinkAdapterV1 changes to use the new Committer and writer creation method, and metrics group (`SinkAdapterV1`, `GlobalCommitterSerializer`, `GlobalCommitterOperator`)
- Updating the tests to propagate the metric group
- Adding a new test case to `SinkV2MetricsITCase` to make sure that the metrics are propagated as expected.

## Verifying this change

Added a new unit test, and modified the old ones to use the new method.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? not documented - might be worth to check the documentation of SinkV2 if it is exists, to update it.

### CODE REVIEW

**Comment by gyfora** [comment]
@reswqa I suggest we focus on reviewing and merging this quickly instead of putting extra work on @pvary .

**Comment by pvary** [comment]
@flinkbot run azure

File: flink-python/pyflink/datastream/connectors/tests/test_pulsar.py [hunk 1]
```diff
@@ -253,7 +250,6 @@ def test_pulsar_sink(self):
                 .long_type()
                 .no_default_value()._j_config_option), 100)
 
-    @unittest.skip("Should be moved to pulsar connector repo")
```
Comment (gyfora) [inline flink-python/pyflink/datastream/connectors/tests/test_pulsar.py:256]: How are all these python changes related to the PR? Should we maybe remove them from here?

Comment (mbalassi) [reply flink-python/pyflink/datastream/connectors/tests/test_pulsar.py:256]: Given the rework there is no need to disable the tests anymore. The previous implementation broke these tests due to the connector externalization work not being fully finished for the Python language binding of the connectors.

**Review by mbalassi — APPROVED** [review]
Looks good to me now after the latest minor improvements. 

@reswqa fair point, @gyfora and I are helping to get this merged as soon as possible. I think the changes are now good, will leave the PR open until next Monday so that others can chime in. If we cannot merge on Monday we should consider splitting into multiple PRs.

File: flink-python/pyflink/datastream/connectors/tests/test_pulsar.py (hunk 1 above)
Comment (pvary) [reply flink-python/pyflink/datastream/connectors/tests/test_pulsar.py:256]: Thanks @mbalassi for answering the question.
This is exactly the situation. This is a revert of disabling the python tests

**Comment by pvary** [comment]
@flinkbot run azure

**Comment by pvary** [comment]
@flinkbot run azure

**Comment by becketqin** [comment]
@pvary Thanks for the patch. I am wondering if you have compared this approach with the alternative I mentioned in the mailing list.

> Option 1:
>1. On top of the the current status, rename org.apache.flink.api.connector.sink2.InitContext to CommonInitContext (should >probably be package private).
>2. Change the name WriterInitContext back to InitContext, and revert the deprecation. We can change the parameter name to writerContext if we want to. 
>Admittedly, this does not have full symmetric naming of the InitContexts - we will have CommonInitContext / InitContext / CommitterInitContext instead of InitContext / WriterInitContext / CommitterInitContext. However, the naming seems clear without much confusion. Personally, I can live with that, treating the class InitContext as a non-ideal legacy class name without much material harm.

The alternative approach has less changes to the existing interface. The only caveat is we have to keep the legacy name of `InitContext`. However, given that we are going to replace `StatefulSink` and `TwoPhaseCommttingSink` with mixin interfaces, the naming issue is just temporary. We can use the new name for InitContexts in the new mixin interfaces.

From this standpoint, personally I prefer the alternative approach as it introduces less changes to the existing interfaces that we are about to deprecate. What do you think?

**Comment by gyfora** [comment]
@becketqin if I understand correctly, we do not need to replace the current interfaces in order to make them mixin. We only need to "deprecate" the inherited methods. 

And the changes that @pvary is introducing affect only those methods, so if we go towards mixin, we might as well fix inconsistent parameter naming in the same step.

**Comment by gyfora** [comment]
The alternative approach would be to deprecate the current interfaces completely (stateful / twophase) and replace them, in which case it is also much better to fix the class structure. No downside from the users perspective as they have to move to new interfaces anyways.

**Comment by pvary** [comment]
> @pvary Thanks for the patch. I am wondering if you have compared this approach with the alternative I mentioned in the mailing list.
[..]
> The alternative approach has less changes to the existing interface. The only caveat is we have to keep the legacy name of `InitContext`. However, given that we are going to replace `StatefulSink` and `TwoPhaseCommttingSink` with mixin interfaces, the naming issue is just temporary. We can use the new name for InitContexts in the new mixin interfaces.
> 
> From this standpoint, personally I prefer the alternative approach as it introduces less changes to the existing interfaces that we are about to deprecate. What do you think?

The typical implementation of the new mixin `Sink` will look like this:
```
public class MyLittleSink implements Sink, WithWriterState, WithCommitter {
    public SinkWriter createWriter(WriterInitContext context) {..}
    public Committer createCommitter(CommitterInitContext context) {..}
}
```

That is why, I think it is better to have different names for different classes.

**Comment by pvary** [comment]
@becketqin: Reverted the rename from this PR. I still think we need to do it, but I will do in the next FLIP/PR combo (FLIP-372)

**Comment by pvary** [comment]
@becketqin: Any more comments? Or do you think it is ready to go?

**Review by gyfora — APPROVED** [review]

**Review by becketqin — APPROVED** [review]
@pvary Thanks for updating the patch and the patience. Looks good to me.

**Review by leonardBang — APPROVED** [review]
Thanks @pvary for the contribution,  a51eff4e2de3f80de61fb45fee81f0de68fd0af3 looks good to me, minor:  the commit id is outdated in PR description .

**Comment by mbalassi** [comment]
Merging manually to preserve the individual revert commits for the sake of the git history while slightly rewording the last commit message.

**Comment by Jiabao-Sun** [comment]
Thanks @pvary for the effort.

I still have a question whether it is necessary for `CommitterInitContext` to inherit `InitContext`. `CommitterInitContext` seems to be independent and only used for creating `Committer`. Can we make it not inherit any interface and still meet the requirements?

```java
    /** The interface exposes some runtime info for creating a {@link Committer}. */
    @PublicEvolving
    interface CommitterInitContext {

        /** @return The id of task where the committer is running. */
        int getSubtaskId();

        /** @return The number of parallel committer tasks. */
        int getNumberOfParallelSubtasks();

        /**
         * Gets the attempt number of this parallel subtask. First attempt is numbered 0.
         *
         * @return Attempt number of the subtask.
         */
        int getAttemptNumber();

        /**
         * Returns id of the restored checkpoint, if state was restored from the snapshot of a previous
         * execution.
         */
        OptionalLong getRestoredCheckpointId();

        /**
         * The ID of the current job. Note that Job ID can change in particular upon manual restart. The
         * returned ID should NOT be used for any job management tasks.
         */
        JobID getJobId();

        /** @return The metric group this committer belongs to. */
        SinkCommitterMetricGroup metricGroup();
    }
```

The compatibility issue for `WriteInitContext` comes from this abstraction. 
Is it possible for us to abandon the abstraction of `InitContext` at the current stage and make further modifications in version 2.0?

**Comment by pvary** [comment]
> Thanks @pvary for the effort.
> 
> I still have a question whether it is necessary for `CommitterInitContext` to inherit `InitContext`. `CommitterInitContext` seems to be independent and only used for creating `Committer`. Can we make it not inherit any interface and still meet the requirements?
[..]
> The compatibility issue for `WriteInitContext` comes from this abstraction. Is it possible for us to abandon the abstraction of `InitContext` at the current stage and make further modifications in version 2.0?

@Jiabao-Sun: If you take a look at [InitContextBase](https://github.com/apache/flink/pull/23876/commits/a51eff4e2de3f80de61fb45fee81f0de68fd0af3#diff-450ec3a61cf8587c258e2ab3d27bb37945126b4d9e6f23560ff5011cb50b0aaf), you can see that quite a bit of code is deduplicated by the inheritance. Also I am not aware of any compatibility issue caused by splitting the original Sink.InitContext interface.

**Comment by Jiabao-Sun** [comment]
Thanks @pvary.

I looked into `InitContextBase`, which mostly wraps the calls to `StreamingRuntimeContext` and does not reduce much redundant code. Instead of the invasiveness to the overall structure, I prefer to define `CommitterInitContext` separately.

**Review by Jiabao-Sun — APPROVED** [review]
However, I still appreciate the compatibility improvements brought about by this PR.
Thanks.

**Comment by gyfora** [comment]
@Jiabao-Sun I think it is better to have a proper class hierarchy for sink related contexts. So I am +1 for what Peter proposed. 

This can help with integration and unified tooling in different parts of the sink. Do you have any technical arguments against the new structure?

**Comment by Jiabao-Sun** [comment]
Hi @gyfora, the proper class hierarchy is good to me and the naming of `WriterInitContext` is reasonable. 
I just think that it is a compromise to rename the `WriterInitContext` to `InitContext`.
Maybe it will be better to go to this change in 2.0.

**Comment by gyfora** [comment]
> Hi @gyfora, the proper class hierarchy is good to me and the naming of `WriterInitContext` is reasonable. I just think that it is a compromise to rename the `WriterInitContext` to `InitContext`. Maybe it will be better to go to this change in 2.0.

Ah sorry, I did not completely understand the comment. We decided to not rename `WriterInitContext` here to avoid any compat issues. But I think we can still introduce the proper hierarchy as that doesn't affect compatibility and allow us to work with the new interface structure.

**Comment by mbalassi** [comment]
Thanks for the review @Jiabao-Sun. I think we reached consensus now for these specific and provided appropriate time for review, so I am proceeding to merge the PR. We can continue the discussion for additional improvements on FLINK-25857 or the mailing list if you have a more elaborate proposal.

**Comment by Jiabao-Sun** [comment]
Thanks @mbalassi, please continue with the merge. 
Just some insignificant thoughts, please ignore it.

> To obtain a proper class hierarchy, we added a new interface `InitContext` and made `CommitterInitContext` and `WriterInitContext` inherit from it. Obviously, in the inheritance structure, it would be more reasonable for `WriterInitContext` to have a symmetric name. However, for compatibility reasons, we had to change `WriterInitContext` back to `InitContext`. This approach will increase the understanding cost for developers as to why there are two `InitContext`s. In addition, the newly added `InitContext` looks very similar to `RuntimeContext`, may also be unnecessary.

**Comment by pvary** [comment]
Thanks @mbalassi for the merge and @gyfora, @reswqa, @becketqin, @leonardBang, @Jiabao-Sun for the reviews

@Jiabao-Sun: While I partially agree with your comments (my original PR #23555 was like that), I think that if we have 2 interfaces where about half of the methods are the same, then we might want to have a common interface and inherit from there. The good thing is that the [FLIP-372](https://cwiki.apache.org/confluence/display/FLINK/FLIP-372%3A+Allow+TwoPhaseCommittingSink+WithPreCommitTopology+to+alter+the+type+of+the+Committable) and the corresponding [discussion](https://lists.apache.org/thread/344pzbrqbbb4w0sfj67km25msp7hxlyd) is still open, so we can continue this debate there.

**Pull request closed without merging**